### PR TITLE
[incubator/etcd] Fixing incubator etcd

### PR DESCRIPTION
**Pull request by Elexy**
This PR:
* fixes rejoining
* adds scaling with `helm upgrade`
* uses arbitrary user (non-root)
* updates the etcd version
* optionally encrypt peer traffic with the built in certs

### CODE REVIEW

File: incubator/etcd/templates/statefulset.yaml [hunk 1]
```diff
@@ -28,147 +28,224 @@ spec:
...
+        #    path: /metrics
+        #    port: {{ .Values.ClientPort }}
         resources:
           requests:
             cpu: "{{ .Values.Cpu }}"
             memory: "{{ .Values.Memory }}"
         env:
         - name: INITIAL_CLUSTER_SIZE
-          value: {{ default 3 .Values.Replicas | quote }}
+          value: {{ default 3 .Values.InitialSize | quote }}
```
Comment (phyrwork) [inline incubator/etcd/templates/statefulset.yaml]: Why do we need an InitialSize var? Why can't we just continue to use the existing var Replicas i.e. current size

File: incubator/etcd/values.yaml [hunk 2]
```diff
@@ -6,13 +6,24 @@
 Name: etcd
 PeerPort: 2380
 ClientPort: 2379
+# ProxyPort: 23790
 Component: "etcd"
+PeerTLS: true
+# InitialSize should be set only at INSTALL, after the cluster is live, don't change this setting.
+InitialSize: 3
```
Comment (phyrwork) [inline incubator/etcd/values.yaml]: Asking for problems by having a setting that shouldn't be changed. As before, why can't we just stick with the one var for current size?

File: incubator/etcd/values.yaml [hunk 3]
```diff
@@ -6,13 +6,24 @@
...
+# This should not be lower that the InitialSize and take note of the following (from etcd project site)
+# The recommended etcd cluster size is 3, 5 or 7, which is decided by the fault tolerance requirement.
+# A 7-member cluster can provide enough fault tolerance in most cases.
+# While larger cluster provides better fault tolerance the write performance reduces since data needs to
+# be replicated to more machines.
 Replicas: 3
-Image: "gcr.io/google_containers/etcd-amd64"
-ImageTag: "2.2.5"
+# ProxyReplicas: 1
+Image: "nearform/docker-etcd"
```
Comment (phyrwork) [inline incubator/etcd/values.yaml]: Non-standard image

File: incubator/etcd/values.yaml [hunk 4]
```diff
@@ -6,13 +6,24 @@
...
+ImageTag: "latest"
```
Comment (phyrwork) [inline incubator/etcd/values.yaml]: Decouples image from chart source. New image could break chart. Should remain as fixed version

File: incubator/etcd/values.yaml [hunk 5]
```diff
@@ -6,13 +6,24 @@
...
 ImagePullPolicy: "Always"
 Cpu: "100m"
-Memory: "512Mi"
+Memory: "256Mi"
```
Comment (phyrwork) [inline incubator/etcd/values.yaml]: Justification for changing default? This is user overridable anyways...

File: incubator/etcd/values.yaml (hunk 3 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: the standard image runs as root, thus insecure.

File: incubator/etcd/values.yaml (hunk 5 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: reading [this](https://coreos.com/etcd/docs/latest/op-guide/hardware.html) doesn't really warrant reducing :-) will revert that change.

File: incubator/etcd/templates/statefulset.yaml (hunk 1 above)
Comment (Elexy) [reply incubator/etcd/templates/statefulset.yaml]: Already removed.

File: incubator/etcd/values.yaml (hunk 2 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: yeah, why I removed it.

File: incubator/etcd/values.yaml (hunk 3 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: The nearform/etcd is an automated build from an [opensource github repo](https://github.com/nearform/docker-etcd) What else can I do, besides urging Coreos to remove the root user dependency. With it Openshift deployments (among others) are not possible.

File: incubator/etcd/values.yaml (hunk 4 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: It was never a fixed image, was previously specified in values.yaml. Perhaps that was the old way of doing charts. I think you are right that a good user experience, a fixed image would be better.

Comment (Elexy) [reply incubator/etcd/values.yaml]: fixed

**Comment by Elexy** [comment]
I commented on all remarks, even the outdated ones. Added two fixes. Anything else from my side required.

File: incubator/etcd/templates/statefulset.yaml [hunk 6]
```diff
@@ -21,7 +21,7 @@ spec:
     spec:
       containers:
       - name: {{ template "etcd.fullname" . }}
-        image: "{{ .Values.Image }}:{{ .Values.ImageTag }}"
+        image: nearform/etcd:3.2.11
```
Comment (alejandroEsc) [inline incubator/etcd/templates/statefulset.yaml]: why are we moving away from values file here?

File: incubator/etcd/templates/statefulset.yaml [hunk 7]
```diff
@@ -34,141 +34,216 @@ spec:
             memory: "{{ .Values.Memory }}"
         env:
         - name: INITIAL_CLUSTER_SIZE
-          value: {{ default 3 .Values.Replicas | quote }}
+          value: "1"
```
Comment (alejandroEsc) [inline incubator/etcd/templates/statefulset.yaml:37]: as above, why explicit value here?

File: incubator/etcd/values.yaml [hunk 8]
```diff
@@ -7,9 +7,13 @@ Name: etcd
...
 Component: "etcd"
+PeerTLS: true
+# Replicas scales the cluster with the "helm upgrade" command
+# The recommended etcd cluster size is 3, 5 or 7, which is decided by the fault tolerance requirement.
+# A 7-member cluster can provide enough fault tolerance in most cases.
+# While larger cluster provides better fault tolerance the write performance reduces since data needs to
+# be replicated to more machines.
 Replicas: 3
-Image: "gcr.io/google_containers/etcd-amd64"
-ImageTag: "2.2.5"
```
Comment (alejandroEsc) [inline incubator/etcd/values.yaml]: again what is the reason for this? what if I wanted to use another image such as my own?

**Review by alejandroEsc — CHANGES_REQUESTED** [review]
just a few questions.

File: incubator/etcd/templates/statefulset.yaml (hunk 6 above)
Comment (Elexy) [reply incubator/etcd/templates/statefulset.yaml]: Changing the image might break the chart. 

@phyrwork  commented 15 days ago
Decouples image from chart source. New image could break chart. Should remain as fixed version

In general I think the chart should be usable with another image as long as there is a description of what is expected of it. I'd be happy to change it back to values.yaml and add a paragraph about the specification of the image in the README.

File: incubator/etcd/templates/statefulset.yaml (hunk 7 above)
Comment (Elexy) [reply incubator/etcd/templates/statefulset.yaml:37]: The mechanism changed, as described in the README.
A cluster is initially 1 node. so if you specify 3 replicas in values.yaml, the first replica forms a 1 node cluster and the remaining 2 are added to the cluster. This allows the cluster to come down to 1 node (for whatever reason) without breaking.

File: incubator/etcd/values.yaml (hunk 8 above)
Comment (Elexy) [reply incubator/etcd/values.yaml]: See my answer above.

File: incubator/etcd/templates/statefulset.yaml (hunk 6 above)
Comment (mattfarina) [reply incubator/etcd/templates/statefulset.yaml]: @Elexy The image should be configurable. It's part of the [review guidelines](https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#configuration). One simple use case is those who copy the image to their own repository rather than using the upstream one.

File: incubator/etcd/README.md [hunk 9]
```diff
@@ -38,15 +38,16 @@ The following tables lists the configurable parameters of the etcd chart and the
 | Parameter               | Description                        | Default                                                    |
 | ----------------------- | ---------------------------------- | ---------------------------------------------------------- |
 | `Name`                  | Spark master name                  | `etcd`                                                     |
-| `Image`                 | Container image name               | `gcr.io/google_containers/etcd-amd64`                      |
-| `ImageTag`              | Container image tag                | `2.2.5`                                                    |
+| `Image`                 | Container image name               | `nearform/docker-etcd`                      |
```
Comment (mattfarina) [inline incubator/etcd/README.md]: I have some questions about the image:

* The image is listed as deprecated. Why should we switch to a deprecated image?
* The image doesn't seem to give a high trust factor. It's a custom image without linking to the source that's manually pushed. Why should we trust the image over the Google provided one?
* What's missing from the current container in use?

**Review by mattfarina — CHANGES_REQUESTED** [review]

File: incubator/etcd/README.md (hunk 9 above)
Comment (Elexy) [reply incubator/etcd/README.md]: Good catch, changed to `nearform/etcd`, an automated build
The original etcd image running as 'root' which is unacceptable for Openshift

**Comment by Elexy** [comment]
@mattfarina changes committed

Bot suggested: 
/assign @seanknox

File: incubator/etcd/README.md [hunk 10]
```diff
@@ -38,15 +38,16 @@ The following tables lists the configurable parameters of the etcd chart and the
 | Parameter               | Description                        | Default                                                    |
 | ----------------------- | ---------------------------------- | ---------------------------------------------------------- |
 | `Name`                  | Spark master name                  | `etcd`                                                     |
-| `Image`                 | Container image name               | `k8s.gcr.io/etcd-amd64`                      |
-| `ImageTag`              | Container image tag                | `2.2.5`                                                    |
+| `Image`                 | Container image name               | `nearform/etcd`                      |
```
Comment (mattfarina) [inline incubator/etcd/README.md:41]: I understand your point on using a non-root user for openshift. Since the image in use today is managed by k8s itself I'd like to chase that down first.

**Comment by hiddeco** [comment]
What is the status?

File: incubator/etcd/README.md (hunk 10 above)
Comment (mattfarina) [reply incubator/etcd/README.md:41]: I was unable to track the person down. Will dig into this as is now.

Comment (Elexy) [reply incubator/etcd/README.md:41]: Is there anything needed from my side? I'll have some time next week.

File: incubator/etcd/templates/statefulset.yaml [hunk 11]
```diff
@@ -34,141 +34,216 @@ spec:
...
+                      rm -rf $ETCD_DIR
                   fi
         command:
           - "/bin/sh"
-          - "-ec"
+          - "-ecx"
           - |
             HOSTNAME=$(hostname)
+            export ETCD_DATA_DIR="/etcd/default.etcd"
+            IP=$(ip addr | grep 'state UP' -A2 | tail -n1 | awk '{print $2}' | cut -f1  -d'/')
```
Comment (hiddeco) [inline incubator/etcd/templates/statefulset.yaml:82]: This does not seem to work for minikube. The `grep` returns no matches resulting in an incomplete ` --listen-client-urls` on line 161.

```bash
$ ip addr
1: lo: &lt;LOOPBACK,UP,LOWER_UP&gt; mtu 65536 qdisc noqueue qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: sit0@NONE: &lt;NOARP&gt; mtu 1480 qdisc noop qlen 1
    link/sit 0.0.0.0 brd 0.0.0.0
530: eth0@if531: &lt;BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN&gt; mtu 1500 qdisc noqueue 
    link/ether 02:42:ac:11:00:0a brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.10/16 scope global eth0
       valid_lft forever preferred_lft forever
```

File: incubator/etcd/README.md [hunk 12]
```diff
@@ -38,15 +38,16 @@ The following tables lists the configurable parameters of the etcd chart and the
...
+| `ImageTag`              | Container image tag                | `2.3.11`                                                    |
```
Comment (hiddeco) [inline incubator/etcd/README.md:42]: `2.3.11` -> `3.2.11`

File: incubator/etcd/README.md (hunk 10 above)
Comment (hiddeco) [inline incubator/etcd/README.md:41]: The image seems to have issues with accessing certain files.

```
2018-01-24 08:25:15.021685 C | embed: could not get certs (mkdir /etcd/default.etcd: permission denied)
```

File: incubator/etcd/Chart.yaml [hunk 13]
```diff
@@ -9,3 +9,5 @@ sources:
 maintainers:
 - name: Lachlan Evenson
   email: lachlan@deis.com
+- name: Alex Knol
```
Comment (mattfarina) [inline incubator/etcd/Chart.yaml:12]: For the maintainer names we now require github usernames. That's so we can more easily contact chart maintainers in their duties at maintainer. Can you please use your github username.

If you feel up to it, you could also update @lachie83 to his username and update or remove his email address as that one no longer works.

**Review by mattfarina — CHANGES_REQUESTED** [review]
@lachie83 since this adds a new maintainer to your chart could you please give it a review.

**Comment by hiddeco** [comment]
Ping

**Comment by rtluckie** [comment]
Tested this patch using k8s.gcr.io/etcd-amd64:3.2.14 everything looks good and we're running it in production.  Also, we've been continually testing with chaoskube to ensure resiliency.
 
I hope this gets merged soon so we can avoid maintaining a fork ;)

**Comment by hiddeco** [comment]
@rtluckie can confirm your fork works, however the advertised members (`http://<service>.<namespace>.svc.cluster.local:2379/v2/machines`) do not provide details about the namespace they're living in.

This makes it hard to combine it with software that makes use of this endpoint for 'auto discovery' ([Patroni](https://github.com/zalando/patroni) for example) where `etcd` lives in a different namespace than the software making use of it as they succeed to list members but fail to connect to the listed hostnames.

I'd propose to use the fully qualified hostname `${PROTO}://${HOSTNAME}.${SET_NAME}.${NAMESPACE}:${PORT}` when adding members to get around this issue.

Patch: https://gist.github.com/hiddeco/f2f6d355f23512214caecc0222caad6a

File: incubator/etcd/values.yaml [hunk 14]
```diff
@@ -21,4 +27,4 @@ Storage: "1Gi"
 ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
 ##   GKE, AWS & OpenStack)
 ##
-# storageClass: "-"
+storageClass: "-"
```
Comment (unguiculus) [inline incubator/etcd/values.yaml:30]: Why is this changed? Our best practice recommends this to be commented out by default.

**Comment by ben-ole** [comment]
hey all, what's the status here?

**Comment by unguiculus** [comment]
ping @Elexy

**Comment by pcornelissen** [comment]
This problem is still there :-(
